Assets/SEE/Dissonance/ChatLogController.cs

```diff
@@ -0,0 +1,327 @@
+            if (textPrototype == null)
+            {
+                Debug.LogError($"Could not find {logTextPrototypePrefab} in Resources.\n");
+                enabled = false;
+                return;
+            }
+
+            if (canvas == null)
+            {
+                Debug.LogError("Canvas not available.\n");
+                enabled = false;
+                return;
+            }
+
+            chatBox = canvas.transform.Find(chatBoxName)?.gameObject;
+            if (chatBox == null)
+            {
+                Debug.LogError($"Could not find chat box named {chatBoxName}.\n");
+                enabled = false;
+                return;
+            }
+            canvasGroup = chatBox.GetComponent<CanvasGroup>();
+
+            heightLimit = chatBox.GetComponent<RectTransform>().rect.height - 20;
+
+            // Subscribe to receive Dissonance comms text messages.
+            if (Comms != null)
+            {
+                Comms.Text.MessageReceived += OnMessageReceived;
+            }
+        }
+
+        /// <summary>
+        /// Prints the <paramref name="message"/> received in the chat log.
+        /// This gets called by Dissonance whenever a text message arrives.
+        /// </summary>
+        /// <param name="message">the message received</param>
+        private void OnMessageReceived(TextMessage message)
+        {
+            // Ignore your own messages coming back from the server.
+            if (Comms != null && message.Sender == Comms.LocalPlayerName)
+            {
+                return;
+            }
+
+            // Decide what we're going to print.
+            string msg = string.Format("{0} ({1}): {2}",
+                message.Sender[..Math.Min(8, message.Sender.Length)],
+                message.RecipientType == ChannelType.Room ? message.Recipient : "Whisper",
+                message.Message
+            );
+
+            AddMessage(msg, new Color(0.19f, 0.19f, 0.19f));
+        }
+
+        /// <summary>
+        /// Adds a message to the chat log.
+        /// </summary>
+        /// <param name="message">the message to be added</param>
+        /// <param name="color">the color in which the message should be printed</param>
+        public void AddMessage(string message, Color color)
+        {
+            EnableCanvas(true);
+
+            // Instantiate the text prototype.
+            GameObject obj = Instantiate(textPrototype, chatBox.transform);
+
+            // Put text into the object.
+            Text txt = obj.GetComponent<Text>();
+            txt.text = message;
+            txt.color = color;
+
+            // Preferred height will be the height it wants to be (e.g. including extra height because of wraparound).
+            // Directly set the height to that height to ensure all text is seen.
+            txt.GetComponent<RectTransform>().sizeDelta = new Vector2(0, txt.preferredHeight);
```

Comment:
RectTransform could be saved in a variable first to avoid the repeated GetComponent calls.

Reply:
Fixed.